modnet/models.py

```diff
@@ -554,79 +566,904 @@ class OR only return the most probable class.
-            pickle.dump(self, f)
-        self.model = model
-        LOG.info("Saved model to {}(.json/.h5/.pkl)".format(filename))
+        model_weights = self.model.get_weights()
+
+        self.model = (model_json, model_weights)
+
+    def _restore_model(self):
+        """
+        restore inner keras model after running make_picklable
+        """
+
+        model_json, model_weights = self.model
+        self.model = tf.keras.models.model_from_json(model_json)
+        self.model.set_weights(model_weights)
+
+
+    def save(self, filename: str):
+        """Save the `MODNetModel` to filename:
+
+        Parameters:
+            filename: The base filename to save to.
+
+        If the filename ends in "tgz", "bz2" or "zip", the pickle
+        will be compressed accordingly by `pandas.to_pickle(...)`.
+
+        """
+        self._make_picklable()
+        pd.to_pickle(self, filename)
+        self._restore_model()
+        LOG.info(f'Model successfully saved as {filename}!')
 
     @staticmethod
     def load(filename: str):
-        """Load the `MODNetModel` from 3 files with the same base
-        filename:
+        """Load `MODNetModel` object pickled by the `.save(...)` method.
 
-        * <filename>.json contains the Keras model JSON dump.
-        * <filename>.pkl contains the `MODNetModel` object, excluding the
-          Keras model.
-        * <filename>.h5 contains the model weights.
+        If the filename ends in "tgz", "bz2" or "zip", the pickle
+        will be decompressed accordingly by `pandas.read_pickle(...)`.
 
         Returns:
             The loaded `MODNetModel` object.
+        """
+        pickled_data = None
+
+        if isinstance(filename, Path):
+            filename = str(filename)
+
+        # handle .zip files explicitly for OS X/macOS compatibility
+        if filename.endswith(".zip"):
+            from zipfile import ZipFile
+            with ZipFile(filename, "r") as zf:
+                namelist = zf.namelist()
+                _files = [_ for _ in namelist if not _.startswith("__MACOSX/") or _.startswith(".DS_STORE")]
+                if len(_files) == 1:
+                    with zf.open(_files.pop()) as f:
+                        pickled_data = pd.read_pickle(f)
+
+        if pickled_data is None:
+            pickled_data = pd.read_pickle(filename)
+
+
+        if isinstance(pickled_data, MODNetModel):
+            if not hasattr(pickled_data, "__modnet_version__"):
+                pickled_data.__modnet_version__ = "unknown"
+            pickled_data._restore_model()
+            LOG.info(f"Loaded {pickled_data} object, created with modnet version {pickled_data.__modnet_version__}")
+            return pickled_data
+
+        raise ValueError(
+            f"File {filename} did not contain compatible data to create a MODNetModel object, "
+            f"instead found {pickled_data.__class__.__name__}."
+        )
+
+def validate_model(train_data = None,
```

Comment:
What does validate mean in this context? If this is just a convenience wrapper to call `fit` in a way that works with multiprocessing, it should be made private and given a more descriptive name (and docstring!)

Reply:
okay